packages/cdk/lib/stacks/context-stack.ts

```diff
@@ -1,31 +1,129 @@
+import { BatchStack, BatchStackProps } from "./nested/batch-stack";
+import { CromwellEngineStack } from "./nested/cromwell-engine-stack";
+import { NextflowEngineStack } from "./nested/nextflow-engine-stack";
 import { ManagedPolicy } from "monocdk/aws-iam";
 
 export interface ContextStackProps extends StackProps {
   readonly contextParameters: ContextAppParameters;
 }
 
 export class ContextStack extends Stack {
+  private readonly vpc: IVpc;
+
   constructor(scope: Construct, id: string, props: ContextStackProps) {
     super(scope, id, props);
 
     const vpcId = getCommonParameter(this, VPC_PARAMETER_NAME);
-    const vpc = Vpc.fromLookup(this, "Vpc", { vpcId });
+    this.vpc = Vpc.fromLookup(this, "Vpc", { vpcId });
+
+    const { contextParameters } = props;
+    const { engineName } = contextParameters;
+
+    switch (engineName) {
+      case "cromwell":
+        this.renderCromwellStack(props);
+        break;
+      case "nextflow":
+        this.renderNextflowStack(props);
+        break;
+      default:
+        throw Error(`Engine '${engineName}' is not supported`);
+    }
+  }
+
+  private renderCromwellStack(props: ContextStackProps) {
+    const batchProps = this.getCromwellBatchProps(props);
+    const batchStack = this.renderBatchStack(batchProps);
+
+    let jobQueue;
+    if (props.contextParameters.requestSpotInstances) {
+      jobQueue = batchStack.batchSpot.jobQueue;
+    } else {
+      jobQueue = batchStack.batchOnDemand.jobQueue;
+    }
+
+    const commonEngineProps = this.getCommonEngineProps(props);
+    new CromwellEngineStack(this, "cromwell", {
+      jobQueue,
+      ...commonEngineProps,
+    }).outputToParent(this);
+  }
 
-    const batchStack = new BatchStack(this, "Batch", { vpc, contextParameters: props.contextParameters });
+  private renderNextflowStack(props: ContextStackProps) {
+    const batchProps = this.getNextflowBatchProps(props);
+    const batchStack = this.renderBatchStack(batchProps);
 
-    const commonProps: CromwellEngineStackProps | NextflowEngineStackProps = {
-      vpc,
+    let jobQueue, headQueue;
+    if (props.contextParameters.requestSpotInstances) {
+      jobQueue = batchStack.batchSpot.jobQueue;
+      headQueue = batchStack.batchOnDemand.jobQueue;
+    } else {
+      headQueue = jobQueue = batchStack.batchOnDemand.jobQueue;
+    }
+
+    const commonEngineProps = this.getCommonEngineProps(props);
+    new NextflowEngineStack(this, "nextflow", {
+      ...commonEngineProps,
+      jobQueue,
+      headQueue,
+    }).outputToParent(this);
+  }
+
+  private getCromwellBatchProps(props: ContextStackProps) {
+    const commonBatchProps = this.getCommonBatchProps(props);
+    if (props.contextParameters.requestSpotInstances) {
+      return {
+        ...commonBatchProps,
+        createSpotBatch: true,
+        createOnDemandBatch: false,
+      };
+    }
+
+    return {
+      ...commonBatchProps,
+      createSpotBatch: false,
+      createOnDemandBatch: true,
+    };
```

Comment:
Could this not be simplified to;

return {
        ...commonBatchProps,
        createSpotBatch: props.contextParameters.requestSpotInstances,
        createOnDemandBatch: true,
      };

Reply:
I think it's a little cleaner to have it this way. I tend to be on "simple is better than smart" part of the spectrum. 

a block like 

```
return {
...commonBatchProps,
createSpotBatch: props.contextParameters.requestSpotInstances,
createOnDemandBatch: !props.contextParameters.requestSpotInstances,,
};
```
could take a second to mentally compile it and figure out what's going on there, where as with the currently proposed code it's pretty simple to read that one case will have only spot and the other would have only on demand.